examples/hierarchical-firewall-policy/main.tf

```diff
@@ -0,0 +1,190 @@
+/**
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+locals {
+  prefix = "hierarchical"
+}
+
+resource "random_string" "random_suffix" {
+  length  = 6
+  special = false
+  lower   = true
+  upper   = false
+}
+
+resource "google_service_account" "service_account" {
+  project      = var.project_id
+  account_id   = "${local.prefix}-fw-test-svc-acct"
+  display_name = "${local.prefix} firewall policy test service account"
+}
+
+resource "google_compute_network" "network" {
+  project = var.project_id
+  name    = "${local.prefix}-network"
+}
+
+resource "google_compute_network" "network_backup" {
+  project = var.project_id
+  name    = "${local.prefix}-network-backup"
+}
+
+module "firewal_policy" {
+  source  = "terraform-google-modules/network/google//modules/hierarchical-firewall-policy"
+  version = "~> 9.0"
+
+  parent_node    = "folders/${var.folder1}"
+  policy_name    = "${local.prefix}-firewall-policy-${random_string.random_suffix.result}"
+  description    = "test ${local.prefix} firewall policy"
+  target_org     = var.org_id
+  target_folders = [var.folder2, var.folder3]
+
+  rules = [
```

Comment:
How are you associating rules to the respective containers? Based on [this diagram](https://cloud.google.com/static/firewall/images/firewall-policies/hfw-1.svg) it seems like rules can be specified at org/folder/project level but this rules block doesn't have that association.

Can you explain that a bit for me?

Reply:
Firewall policy itself can attach to org or or folder. 
Rules can be for specific VMs with service account which are provided as `target_service_accounts` or you can apply a rule to a specific VPC which is provided as `target_resources`. These are the only options available for hierarchical firewall policies

Reply:
I see. So line 51 and 52 are default targets and additionally a `target_resources` slice in `rules` adds an additional FW rule for the for those resources specifically. Did I get that right?

Reply:
Line 51 and 52 applies whole firewall policy (all rules) to folders or an org. But Line 82, 103, 143 and 163 allow users to apply a firewall rule to VMs of specific VPC network of VM with specific service account instead of all the VMs under the folders/org.